client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java

```diff
@@ -63,66 +74,110 @@ public Segment getSegmentForKey(String key) {
+        Segment last = null;
+        for (Iterator<SegmentWithRange> iterator = result.descendingMap().values().iterator(); iterator.hasNext();) {
+            SegmentWithRange current = iterator.next();
+            if (current.getSegment().equals(last)) {
+                iterator.remove();
+            }
+            last = current.getSegment();
+        }
+    }
+
+    private SegmentWithRange findReplacedSegment(Segment segment) {
+        return segments.values()
+                       .stream()
+                       .filter(withRange -> withRange.getSegment().equals(segment))
+                       .findFirst()
+                       .orElseThrow(() -> new IllegalArgumentException("Segment to be replaced should be present in the segment list"));
+    }
+    
     /**
      * Checks that replacementSegments provided are consistent with the segments that currently being used.
+     * @param replacedSegment The segment on which EOS was reached
      * @param replacementSegments The StreamSegmentsWithPredecessors to verify
      */
-    private void verifyReplacementRange(StreamSegmentsWithPredecessors replacementSegments) {
+    private void verifyReplacementRange(SegmentWithRange replacedSegment, StreamSegmentsWithPredecessors replacementSegments) {
         log.debug("Verification of replacement segments {} with the current segments {}", replacementSegments, segments);
         Map<Long, List<SegmentWithRange>> replacementRanges = replacementSegments.getReplacementRanges();
+        List<SegmentWithRange> replacements = replacementRanges.get(replacedSegment.getSegment().getSegmentId());
+        Preconditions.checkArgument(replacements != null, "Replacement segments did not contain replacements for segment being replaced");
+        if (replacementRanges.size() == 1) {
+            //Simple split
+            Preconditions.checkArgument(replacedSegment.getHigh() == getUpperBound(replacements));
+            Preconditions.checkArgument(replacedSegment.getLow() == getLowerBound(replacements));
+        } else {
+            Preconditions.checkArgument(replacedSegment.getHigh() <= getUpperBound(replacements));
+            Preconditions.checkArgument(replacedSegment.getLow() >= getLowerBound(replacements));
+        }
         for (Entry<Long, List<SegmentWithRange>> ranges : replacementRanges.entrySet()) {
-            double lowerReplacementRange = 1;
-            double upperReplacementRange = 0;
-            for (SegmentWithRange range : ranges.getValue()) {
-                upperReplacementRange = Math.max(upperReplacementRange, range.getHigh());
-                lowerReplacementRange = Math.min(lowerReplacementRange, range.getLow());
-            }
-            Entry<Double, Segment> upperReplacedSegment = segments.floorEntry(upperReplacementRange);
-            Entry<Double, Segment> lowerReplacedSegment =  segments.higherEntry(lowerReplacementRange);
+            double upperReplacementRange = getUpperBound(ranges.getValue());
```

Comment:
This is just a minor comment, but the variable is a single value, so not really a range.

Reply:
Documented